### ISSUE TITLE
Backpropogation error with tf.math.top_k

### ISSUE THREAD

**Issue by tokotchd** [tensorflow/tensorflow#29971]
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):  
No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Ubuntu 18.04
- TensorFlow installed from (source or binary):
Source
- TensorFlow version (use command below):
1.13
- Python version:
3.5.7 (Also reproduced with 3.6)
- CUDA/cuDNN version:
10.1
- GPU model and memory:
Quadro P3000

**Describe the current behavior**
When the input argument in tf.math.top_k losses is a tensor of shape [0, 0], backpropogation fails on the reshape step. 

**Describe the expected behavior**
There should be no backprop on collecting 0 elements from tensor with 0 values. 

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
Trying to perform Online Hard Example Mining with two lists of losses of shape [None, None] (Batch, num_losses)
When there are either zero positive examples or zero negative examples, trips the error on the backprop step even though there should be no backprop for a selection of zero elements or a tensor of zero elements
```
(placeholders/ setup)
positive_class_examples = tf.gather_nd(classification_losses, positive_class_indices)
negative_class_examples = tf.gather_nd(classification_losses, negative_class_indices)
num_pos_examples = tf.shape(positive_class_examples)[0]
num_neg_examples = tf.math.maximum(256 - num_pos_examples, 0)
top_negative_losses, top_negative_loss_indices = tf.math.top_k(negative_class_examples, k=num_neg_examples)
balanced_top_classification_losses = tf.concat([positive_class_examples, top_negative_losses], axis=-1)
...
(optimizer creation with prior code included)
```

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.
```
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1334, in _do_call
    return fn(*args)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1319, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1407, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Reshape cannot infer the missing input size for an empty tensor unless all specified input sizes are non-zero
	 [[{{node gradients/balance_positive_and_negative_examples/TopKV2_grad/Reshape}}]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  <Redacted>
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 929, in run
    run_metadata_ptr)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1152, in _run
    feed_dict_tensor, options, run_metadata)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1328, in _do_run
    run_metadata)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1348, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Reshape cannot infer the missing input size for an empty tensor unless all specified input sizes are non-zero
	 [[node gradients/balance_positive_and_negative_examples/TopKV2_grad/Reshape (defined at train.py:177) ]]

Caused by op 'gradients/balance_positive_and_negative_examples/TopKV2_grad/Reshape', defined at:
  <Redacted>
    train_op = optimizer.minimize(loss)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/training/optimizer.py", line 403, in minimize
    grad_loss=grad_loss)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/training/optimizer.py", line 512, in compute_gradients
    colocate_gradients_with_ops=colocate_gradients_with_ops)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 664, in gradients
    unconnected_gradients)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 965, in _GradientsHelper
    lambda: grad_fn(op, *out_grads))
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 420, in _MaybeCompile
    return grad_fn()  # Exit early
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 965, in <lambda>
    lambda: grad_fn(op, *out_grads))
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/nn_grad.py", line 1002, in _TopKGrad
    ind_2d = array_ops.reshape(op.outputs[1], array_ops.stack([-1, ind_lastdim]))
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/gen_array_ops.py", line 7179, in reshape
    "Reshape", tensor=tensor, shape=shape, name=name)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3300, in create_op
    op_def=op_def)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1801, in __init__
    self._traceback = tf_stack.extract_stack()

...which was originally created as op 'balance_positive_and_negative_examples/TopKV2', defined at:
  <Redacted>
  File "train.py", line 127, in balance_positive_and_negative_examples
    top_negative_losses, top_negative_loss_indices = tf.math.top_k(negative_class_examples, k=num_neg_examples)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/nn_ops.py", line 3084, in top_k
    return gen_nn_ops.top_kv2(input, k=k, sorted=sorted, name=name)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/gen_nn_ops.py", line 8401, in top_kv2
    "TopKV2", input=input, k=k, sorted=sorted, name=name)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3300, in create_op
    op_def=op_def)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1801, in __init__
    self._traceback = tf_stack.extract_stack()

InvalidArgumentError (see above for traceback): Reshape cannot infer the missing input size for an empty tensor unless all specified input sizes are non-zero
	 [[node gradients/balance_positive_and_negative_examples/TopKV2_grad/Reshape (defined at train.py:177) ]]
```

**Comment by gadagashwini** [comment]
@tokotchd In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by tokotchd** [comment]
```import tensorflow as tf
import numpy as np

input_placeholder = tf.placeholder(dtype=tf.float32, shape=[None, 48, 48, 3])
target_for_optimizer = tf.placeholder(dtype=tf.float32, shape=[None, 48, 48, 3])
num_positive_examples_chosen = tf.placeholder(dtype=tf.int32, shape=None)

junk_conv_layers = tf.layers.conv2d(input_placeholder, filters=3, strides=[1,1], kernel_size=[1,1])

l1_diff = tf.abs(junk_conv_layers - target_for_optimizer)
losses = tf.reduce_mean(l1_diff, axis=3)
losses = tf.reduce_mean(losses, axis=2)
losses = tf.reduce_mean(losses, axis=1)

# positive_class_examples = tf.gather_nd(classification_losses, positive_class_indices)
# negative_class_examples = tf.gather_nd(classification_losses, negative_class_indices)
# num_pos_examples = tf.shape(positive_class_examples)[0]
# num_neg_examples = tf.math.maximum(256 - num_pos_examples, 0)
# top_negative_losses, top_negative_loss_indices = tf.math.top_k(negative_class_examples, k=num_neg_examples)
# balanced_top_classification_losses = tf.concat([positive_class_examples, top_negative_losses], axis=-1)

top_loss_picks, top_pick_indices = tf.math.top_k(losses, k=num_positive_examples_chosen)

optimizer = tf.train.AdamOptimizer(learning_rate=0.00001)
train_op = optimizer.minimize(top_loss_picks)

with tf.Session() as session:
    session.run(tf.initialize_all_variables())
    random_inputs = np.random.rand(20, 48, 48, 3)
    random_target = np.random.rand(20, 48, 48, 3)
    number_of_top_losses = np.random.randint(0, 2, size=[])
    _ = session.run(train_op, feed_dict={input_placeholder: random_inputs, target_for_optimizer: random_target, num_positive_examples_chosen: number_of_top_losses})
```

After building a minimum working example, it seems that the only condition required for this error is backpropogation on top_k function when k == 0.  As a result, the above code runs successfully 50% of the time and fails with the following stacktrace the other 50% of the time.
```
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1334, in _do_call
    return fn(*args)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1319, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1407, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Reshape cannot infer the missing input size for an empty tensor unless all specified input sizes are non-zero
	 [[{{node gradients/TopKV2_grad/Reshape}}]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "minimum_example.py", line 32, in <module>
    _ = session.run(train_op, feed_dict={input_placeholder: random_inputs, target_for_optimizer: random_target, num_positive_examples_chosen: number_of_top_losses})
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 929, in run
    run_metadata_ptr)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1152, in _run
    feed_dict_tensor, options, run_metadata)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1328, in _do_run
    run_metadata)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1348, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Reshape cannot infer the missing input size for an empty tensor unless all specified input sizes are non-zero
	 [[node gradients/TopKV2_grad/Reshape (defined at minimum_example.py:25) ]]

Caused by op 'gradients/TopKV2_grad/Reshape', defined at:
  File "minimum_example.py", line 25, in <module>
    train_op = optimizer.minimize(top_loss_picks)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/training/optimizer.py", line 403, in minimize
    grad_loss=grad_loss)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/training/optimizer.py", line 512, in compute_gradients
    colocate_gradients_with_ops=colocate_gradients_with_ops)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 664, in gradients
    unconnected_gradients)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 965, in _GradientsHelper
    lambda: grad_fn(op, *out_grads))
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 420, in _MaybeCompile
    return grad_fn()  # Exit early
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 965, in <lambda>
    lambda: grad_fn(op, *out_grads))
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/nn_grad.py", line 1002, in _TopKGrad
    ind_2d = array_ops.reshape(op.outputs[1], array_ops.stack([-1, ind_lastdim]))
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/gen_array_ops.py", line 7179, in reshape
    "Reshape", tensor=tensor, shape=shape, name=name)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3300, in create_op
    op_def=op_def)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1801, in __init__
    self._traceback = tf_stack.extract_stack()

...which was originally created as op 'TopKV2', defined at:
  File "minimum_example.py", line 22, in <module>
    top_loss_picks, top_pick_indices = tf.math.top_k(losses, k=num_positive_examples_chosen)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/nn_ops.py", line 3084, in top_k
    return gen_nn_ops.top_kv2(input, k=k, sorted=sorted, name=name)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/ops/gen_nn_ops.py", line 8401, in top_kv2
    "TopKV2", input=input, k=k, sorted=sorted, name=name)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3300, in create_op
    op_def=op_def)
  File "/home/user/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1801, in __init__
    self._traceback = tf_stack.extract_stack()

InvalidArgumentError (see above for traceback): Reshape cannot infer the missing input size for an empty tensor unless all specified input sizes are non-zero
	 [[node gradients/TopKV2_grad/Reshape (defined at minimum_example.py:25) ]]
```

**Comment by gadagashwini** [comment]
I could reproduce the issue on colab with TF-gpu 1.13.1.

**Comment by alextp** [comment]
I cannot reproduce this using nightly.

Can you find a smaller reproducing example? Say, just the shape of the input to top_k, a call to top_k, and a call to gradients?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tokotchd** [comment]
@alextp you need some sort of trainable variables/layers otherwise the call to gradients will not propagate through top_k.

**Comment by namedysx** [comment]
I had the same problem.
how to solve this problem

**Comment by namedysx** [comment]
@alextp  

I ran his code three times, and I get this error.
maybe you should run the code he provided a few times to get this error.

**Comment by gadagashwini** [comment]
@tokotchd This is fixed in Tf 1.14.0. Please take a look at colab gist [here](https://colab.research.google.com/drive/1Q7HkreVszvG6ygXqlQrUoZI0ZRHzsX4m). You want to give a try. Thanks!

**Comment by alextp** [comment]
I just ran this 20 times with and without GPUs on colab and did not get the error.
